### PR TITLE
Fix sqlite memory connection issues

### LINKED ISSUE #2073: sqlite tests failing in Fedora packaging environment
I am lost on this issue. When running the tests, I get the following error:
```console
---- database::test::test_search_bench_dupes stdout ----
thread 'database::test::test_search_bench_dupes' panicked at atuin-client/src/database.rs:1069:18:
called `Result::unwrap()` on an `Err` value: Database(SqliteError { code: 1, message: "no such table: history" })
```

Running locally I am not able to reproduce this error.

I have created a mock environment (equivalent to a docker container but for package builds):
<details><summary>Mock environment</summary>

`~/.config/mock/atuin.cfg`
```cfg
include('/etc/mock/fedora-rawhide-x86_64.cfg')

config_opts['root'] = 'fedora-rawhide-atuin'

config_opts[f'{config_opts.package_manager}.conf'] += """

[atuin]
name=Atuin copr
baseurl=https://copr-be.cloud.fedoraproject.org/results/packit/SriRamanujam-atuin-rpmspec-7/fedora-rawhide-$basearch/
type=rpm-md
skip_if_unavailable=False
gpgcheck=1
gpgkey=https://copr-be.cloud.fedoraproject.org/results/packit/SriRamanujam-atuin-rpmspec-7/pubkey.gpg
repo_gpgcheck=0
enabled=1
enabled_metadata=1
priority=9

[paserk_paseto]
name=Paserk paseto copr
baseurl=https://copr-be.cloud.fedoraproject.org/results/packit/LecrisUT-rusty_parserk-paseto-rpmspec-main/fedora-rawhide-$basearch/
type=rpm-md
skip_if_unavailable=False
gpgcheck=1
gpgkey=https://copr-be.cloud.fedoraproject.org/results/packit/LecrisUT-rusty_parserk-paseto-rpmspec-main/pubkey.gpg
repo_gpgcheck=0
enabled=1
enabled_metadata=1
priority=9

[metrics]
name=Metrics copr
baseurl=https://copr-be.cloud.fedoraproject.org/results/packit/LecrisUT-rust-metris-rpmspec-main/fedora-rawhide-$basearch/
type=rpm-md
skip_if_unavailable=False
gpgcheck=1
gpgkey=https://copr-be.cloud.fedoraproject.org/results/packit/LecrisUT-rust-metris-rpmspec-main/pubkey.gpg
repo_gpgcheck=0
enabled=1
enabled_metadata=1
priority=9

[sqlx]
name=Sqlx copr
baseurl=https://copr-be.cloud.fedoraproject.org/results/packit/LecrisUT-sqlx-rpmspec-main/fedora-rawhide-$basearch/
type=rpm-md
skip_if_unavailable=False
gpgcheck=1
gpgkey=https://copr-be.cloud.fedoraproject.org/results/packit/LecrisUT-sqlx-rpmspec-main/pubkey.gpg
repo_gpgcheck=0
enabled=1
enabled_metadata=1
priority=9
"""
```

And running the following to get into the build environment (don't know how reproducible this is in other environments)
```console
$ wget https://download.copr.fedorainfracloud.org/results/packit/SriRamanujam-atuin-rpmspec-7/fedora-rawhide-x86_64/07513006-atuin/atuin-18.2.0-1.fc41.src.rpm
^Downloads the srpm to get the source and try build atuin^
$ mock -r atuin ./atuin-18.2.0-1.fc41.src.rpm
^Runs the build process^
$ mock -r atuin --install emacs less
^Install some tools you may need^
$ mock -r atuin --shell
^Enter the build environment^
^Next is just some setups to mimic the build environment^
# cd /builddir/build/BUILD/atuin-18.2.0-build/atuin-18.2.0
# export CARGO_HOME=/builddir/build/BUILD/atuin-18.2.0-build/atuin-18.2.0/.cargo
# export RUSTFLAGS="-Copt-level=3 -Cdebuginfo=2 -Ccodegen-units=1 -Cstrip=none -Cforce-frame-pointers=yes -Clink-arg=-specs=/usr/lib/rpm/redhat/redhat-package-notes --cap-lints=warn"
# export RUSTC_BOOTSTRAP=1
# export RPM_PACKAGE_NAME=atuin
# export RPM_PACKAGE_VERSION=18.2.0
# export RPM_PACKAGE_RELEASE=1
^Next I just run a random test^
# cargo test -Z avoid-dev-deps --profile rpm --package atuin-client --lib database::test::test_search_fuzzy -- --exact
```

</details>

From all of my experimentation with `debug` vs `--profile rpm`, removing `CARGO_HOME` and `RUSTFLAGS`, removing `-Z avoid-dev-deps`,  and such, I couldn't find something in there that caused  such issues.

I have tried editing `atuin-client/migrations` to make it break, and indeed it showed that the migration was actually run, so I am not sure what else to try for this issue.

### PR DESCRIPTION
**Pull request by LecrisUT**
Basically for in-memory sqlite databases it can happen that the pool sharing is not handled properly and the database overwrites the old one. Since the in-memory databases are used only for testing it should be fine to simply ensure the connection lives forever. Thoughts?

Resolves #2073
Reference:
- https://github.com/launchbadge/sqlx/issues/2510
- https://github.com/launchbadge/sqlx/issues/362#issuecomment-636661146

## Checks
- [x] I am happy for maintainers to push small adjustments to this PR, to speed up the review cycle
- [x] I have checked that there are no existing pull requests for the same thing

### CODE REVIEW

File: crates/atuin-client/src/database.rs [hunk 1]
```diff
@@ -144,10 +144,21 @@ impl Sqlite {
...
-            .connect_with(opts)
-            .await?;
+        let mut pool_opts = SqlitePoolOptions::new()
+            .acquire_timeout(Duration::from_secs_f64(timeout));
+
+        // Workaround for sqlx-sqlite in-memory failure.
+        // Make sure a single connection is created to avoid overwriting the database.
+        // Safe to use in this case since memory databases are only used for testing.
+        // https://github.com/launchbadge/sqlx/issues/2510
+        if path.to_str().unwrap().ends_with(":memory:"){
```
Comment (LecrisUT) [inline crates/atuin-client/src/database.rs:154]: I wanted to use [`opts.in_memory()`](https://github.com/launchbadge/sqlx/pull/3137) but that will only come in `sqlx 0.8

**Comment by ellie** [comment]
Hey!

Do you have an example run in our CI where this fails? 

According to this comment, below the one you linked, this issue should be fixed in SQLx: https://github.com/launchbadge/sqlx/issues/362#issuecomment-641877214

I've just ran our unit tests 100x both on mac and linux, and on neither platform can I replicate a failure.

**Comment by LecrisUT** [comment]
Not in the local CI because the system `sqlite` is not used, but this is reproduced with `libsqlite3-sys` package on Fedora. I believe that's the case because the original issue is not reproduced with the vendored packages (https://github.com/SriRamanujam/atuin-rpmspec/pull/11 vs https://github.com/SriRamanujam/atuin-rpmspec/pull/7).

> According to this comment, below the one you linked, this issue should be fixed in SQLx

Yes this is an upstream issue. The fix is relatively simple to incorporate here though until these fixes are found over there.

**Review by ellie — CHANGES_REQUESTED** [review]
A couple of comments, though I'm not sure this change is needed.

EG, this approach suggests using a named in memory file: https://github.com/launchbadge/sqlx/issues/2510#issuecomment-1750535623

Which it looks like they already do: https://github.com/launchbadge/sqlx/blob/1388fc8acc07b41ab38f4dc096f03ba9c7b851f8/sqlx-sqlite/src/options/parse.rs#L18-L22

**Comment by ellie** [comment]
> Not in the local CI because the system sqlite is not used, **but this is reproduced with libsqlite3-sys package on Fedora.** 

I'm afraid we do not support using the system sqlite. If you can replicate the failure with our lockfile, then I'm happy to merge the fixes. Please don't upstream any more changes unless they are fixing a failure that can be replicated with our build.

**Comment by ellie** [comment]
Also, I know I linked you it before, but it looks like you need to enable `SQLITE_USE_URI` in your version of sqlite.

Void, having the same issue: https://forum.atuin.sh/t/packaging-atuin-for-void-linux/20/6?u=ellie
Void, finding the solution: https://forum.atuin.sh/t/packaging-atuin-for-void-linux/20/9?u=ellie
Your version, not having the flag set: https://src.fedoraproject.org/rpms/sqlite/blob/rawhide/f/sqlite.spec#_145

**Pull request closed without merging**